portsyncd/linksync.cpp

```diff
@@ -25,63 +32,114 @@ LinkSync::LinkSync(DBConnector *db) :
 {
-    if ((nlmsg_type != RTM_NEWLINK) && (nlmsg_type != RTM_GETLINK) &&
-        (nlmsg_type != RTM_DELLINK))
+    if ((nlmsg_type != RTM_NEWLINK) && (nlmsg_type != RTM_DELLINK))
         return;
 
     struct rtnl_link *link = (struct rtnl_link *)obj;
     string key = rtnl_link_get_name(link);
-    if (nlmsg_type == RTM_DELLINK) /* Will be sync by other application */
+
+    if (key == "lo" || key == "eth0" || key == "docker0" || key == "bcm0")
         return;
 
-    bool admin_state = rtnl_link_get_flags(link) & IFF_UP;
-    bool oper_state = rtnl_link_get_flags(link) & IFF_LOWER_UP;
+    unsigned int flags = rtnl_link_get_flags(link);
+    bool admin_state = flags & IFF_UP;
+    bool oper_state = flags & IFF_LOWER_UP;
     unsigned int mtu = rtnl_link_get_mtu(link);
 
+    char addrStr[MAX_ADDR_SIZE+1] = {0};
+    nl_addr2str(rtnl_link_get_addr(link), addrStr, MAX_ADDR_SIZE);
+
+    unsigned int ifindex = rtnl_link_get_ifindex(link);
+    int master = rtnl_link_get_master(link);
+    char *type = rtnl_link_get_type(link);
+
+    cout << "Receive nlmsg from portsyncd: type:" << nlmsg_type << " key:" << key
+         << " admin_state:" << admin_state << " oper_state:" << oper_state
+         << " addr:" << addrStr << " ifindex:" << ifindex << " master:" << master;
+    if (type)
+        cout << " type:" << type;
+    cout << endl;
+
+    /* Insert or update the ifindex to key map */
+    m_ifindexNameMap[ifindex] = key;
+
+    /* Will be dealt by teamsyncd */
+    if (type && !strcmp(type, TEAM_DRV_NAME))
+        return;
+
     vector<FieldValueTuple> fvVector;
+
+    /* VLAN member: A separate entry in VLAN_TABLE will be inserted */
+    if (master)
+    {
+        key = m_ifindexNameMap[master] + ":" + key;
+
+        if (nlmsg_type == RTM_DELLINK)
+            m_vlanTableProducer.del(key);
+        else
+        {
+            FieldValueTuple t("tagging_mode", "untagged");
+            fvVector.push_back(t);
+
+            m_vlanTableProducer.set(key, fvVector);
+        }
+    }
+
     FieldValueTuple a("admin_status", admin_state ? "up" : "down");
     FieldValueTuple o("oper_status", oper_state ? "up" : "down");
     FieldValueTuple m("mtu", to_string(mtu));
     fvVector.push_back(a);
     fvVector.push_back(o);
     fvVector.push_back(m);
 
+    /* VLAN interfaces: Check if the type is bridge */
+    if (type && !strcmp(type, VLAN_DRV_NAME))
+    {
+        if (nlmsg_type == RTM_DELLINK)
+            m_vlanTableProducer.del(key);
+        else
+            m_vlanTableProducer.set(key, fvVector);
+
+        return;
+    }
+
+    /* front panel interfaces: Check if the port is in the PORT_TABLE */
     vector<FieldValueTuple> temp;
-    if (m_lagTableConsumer.get(key, temp))
-        m_lagTableProducer.set(key, fvVector);
-    else if (m_vlanTableConsumer.get(key, temp))
-        m_vlanTableProducer.set(key, fvVector);
-    else if (m_portTableConsumer.get(key, temp))
+    if (m_portTableConsumer.get(key, temp))
     {
```

Comment:
lag will be dealt in the above code, i'll add logs for the entries that cannot be found in the port table.